### ISSUE TITLE
Raise exceptions that are compatible with Python version

### ISSUE THREAD

**Issue by wbyoung** [celery/celery#2920]
I've just started using this package in a project I'm working on, but noticed that [this line](https://github.com/celery/celery/blob/a19f080/celery/result.py#L176) is not raising an actual exception and made it a little harder to track down an error. I forget what version of Python disallowed exceptions being raised that weren't inheriting from `BaseException`, but I think it's been a while.

Is there anything that would prevent these from being wrapped into an actual exception object? Backwards compatibility perhaps?

There are a few other cases in this file, and perhaps in the code base outside of this file.

- [L274](https://github.com/celery/celery/blob/a19f080/celery/result.py#L274)
- [L861](https://github.com/celery/celery/blob/a19f080/celery/result.py#L861)
- [L702](https://github.com/celery/celery/blob/a19f080/celery/result.py#L702)

If this is for backwards compatibility, would you accept a PR that raised a proper exception on Python versions that require it? Something like:

```python
import sys

def raise_generic(value):
    # assuming 2.4 was the last version that allowed
    # raising any value, but I'd confirm the actual version
    if sys.version_info[0] > 2 or sys.version_info[1] > 4:
        raise Exception(value)
    else:
        raise value
````

**Comment by ask** [comment]
It's raising the serialized exceptions, if it's anything but an exception then it's a bug in the result backend.

**Comment by wbyoung** [comment]
@ask thanks for the quick response.

It looks an awful lot like the result backend is throwing an exception. The object that's being returned is:

```python
{
  'status': 'FAILURE',
  'traceback': '...',
  'result': {'exc_message': "u'first_name'", 'exc_type': 'KeyError'},
  'task_id': 'a_task_id',
  'children': []
}
```

And the code in the result backend that's being pointed to is basically just:

```python
try:
   object['first_name']
except Exception, e:
   raise e
```

I'll look into it on my end for sure, but regardless, wouldn't it make sense to ensure what's being raised is an exception?

For instance, if I was working with a 3rd party service that made a mistake, it'd be nice to get an error message back from Celery telling me that's the case & includes the details from `meta['result']` rather than the useless python message.

**Comment by ask** [comment]
If it's trying to raise the dict, then it means that this particular result backend didn't convert the dict representation back into a Python exception using `Backend.meta_from_decoded(meta)`.

**Comment by ask** [comment]
Btw, please include the result backend you are using, and also the full traceback!

**Comment by wbyoung** [comment]
@ask it's a custom backend — sorry I can't give more details or point to a specific backend. I updated the comment above to include the traceback. As I mentioned, the backend is really just re-rasing & appears to be doing so for a simple `KeyError`.

**Comment by ask** [comment]
Doesn't look like a KeyError? Seems to be raising a json serialized exception dict, with exc_type keys and all.

**Comment by wbyoung** [comment]
@ask yeah, I can't be 100% certain, but I traced it through the code base and it appeared to be an exception.

Regardless, the point that I was trying to make earlier is that it doesn't really matter if it's a bug in the backend. At the point that the "exception" is being raised in the lines of code referenced in the original comment, `celery` can't be sure that it's raising an exception. It's just raising whatever the result backend sent to it. I understand that it's a bug in the backend, but it's also a "bug" in `celery` since you can't be sure that what you're raising is an exception.

Here's why I originally opened this issue. Debugging this is really not fun:

```
Traceback (most recent call last):
  File "./client.py", line 22, in run
    data = request.get(timeout=timeout)
  File "/home/wyoung/dev/api/env/lib/python3.3/site-packages/celery/result.py", line 176, in get
    raise meta['result']
TypeError: exceptions must derive from BaseException
```

Assuming everything goes right, `celery` raises an exception. But when does everything go right? Ensuring that `celery` raises an exception rather than a dict would certainly be helpful in tracking down these types of issues. You could raise with a message saying _the result backed raised something other than an exception which is an error. It raised: blah blah repr of thing it raised_. That'd be really helpful when things don't go right.

**Comment by wbyoung** [comment]
This was due to running an out of date version of `celery` and was resolved by the same solution as #2441, the change 39592832f2ff8623bfa9814cdf4a42c71d5cc222. So @ask, you were right with the suggestion that it wasn't deserializing via `Backend.meta_from_decoded(meta)`.
